Checked LINQ queries and collections for inefficiencies.

RoosterDragon opened the pull request:
- Made Array.IndexOf available via extension method.
- Made ToHashSet extension method.
- Change collections queried often via Contains into sets.
- Avoid Count() extension if Count or Length property exist.
- Made Count() > 0 checks and variations calls to Any() instead.
- Don't call ToList/ToArray if there is no benefit to materializing the sequence.
- If the sequence does benefit from materialization, follow this general pattern:
  - Collection queried often via Contains use ToHashSet to speed up lookups.
  - Short lived variables use ToList. This is because ToArray requires an extra copy to output the final size.
  - Collections persisted into fields or for a long time use ToArray to minimize memory overhead.

RoosterDragon committed:
Checked LINQ queries and collections for inefficiencies.

- Made Array.IndexOf available via extension method.
- Made ToHashSet extension method.
- Change collections queried often via Contains into sets.
- Avoid Count() extension if Count or Length property exist.
- Made Count() > 0 checks and variations calls to Any() instead.
- Don't call ToList/ToArray if there is no benefit to materializing the sequence.
- If the sequence does benefit from materialization, follow this general pattern:
  - Collection queried often via Contains use ToHashSet to speed up lookups.
  - Short lived variables use ToList. This is because ToArray requires an extra copy to output the final size.
  - Collections persisted into fields or for a long time use ToArray to minimize memory overhead.

RoosterDragon commented:
This is a somewhat old set of changes I had thought tested but clearly I haven't. I'll have a more through go over of them for a bit as well as addressing the issues noted.

penev92 commented on OpenRA.Game/Effects/FlashTarget.cs:
```diff
@@ -25,7 +25,7 @@ public FlashTarget(Actor target, Player asPlayer = null, int ticks = 4)
 			player = asPlayer;
 			remainingTicks = ticks;
-			foreach (var e in target.World.Effects.OfType<FlashTarget>().Where(a => a.target == target).ToArray())
+			foreach (var e in target.World.Effects.OfType<FlashTarget>().Where(a => a.target == target).ToList())
```
>Short lived variables use ToList. This is because ToArray requires an extra copy to output the final size.

My question is why explicitly enumerate this at all?

ScottNZ replied:
You need to take a copy (using either ToArray or ToList) and enumerate over that instead if you want to remove items while you are enumerating over it with foreach. This avoids a sanity-check exception which would otherwise be thrown.

RoosterDragon replied:
In this case, because the collection size is unknown, ToList is very slightly faster than ToArray.

ScottNZ replied:
Sounds a bit like premature optimisation then.

RoosterDragon replied:
Probably.

RoosterDragon commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -71,12 +71,11 @@ List<IRenderable> GenerateRenderables()
 			// Include player actor for the rendered player
 			if (world.RenderPlayer != null)
-				actors.Add(world.RenderPlayer.PlayerActor);
+				actors.Append(world.RenderPlayer.PlayerActor);
```
Indeed that'll be the reason for the missing trees I should expect.

RoosterDragon commented:
Ok I played a few games and tested a few other things that may have been affected (like selection and the replay widget) and didn't notice any additional problems beyond the now fixed trees thing.

Mailaender commented:
Needs a rebase now.

RoosterDragon commented:
Rebased.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -26,7 +26,7 @@ public class World
 		static readonly Func<int, int, bool> FalsePredicate = (u, v) => false;
 		internal readonly TraitDictionary TraitDict = new TraitDictionary();
-		readonly HashSet<Actor> actors = new HashSet<Actor>();
+		readonly List<Actor> actors = new List<Actor>();
```
What is the advantage of this?

RoosterDragon replied:
In the SyncHash method, the actors and effects in enumerated to generate the hash, and the order of enumeration is important in this method. Since HashSet has no defined order, this is undefined behaviour.

We need collections with a defined order, so I just went with List to match the existing effects collection.

Thinking about it though, SortedSet makes more sense - might change that.

pchote commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -71,12 +71,11 @@ List<IRenderable> GenerateRenderables()
 			// Include player actor for the rendered player
 			if (World.RenderPlayer != null)
-				actors.Add(World.RenderPlayer.PlayerActor);
+				actors = actors.Append(World.RenderPlayer.PlayerActor);
```
Maybe use `AddRange`, like you do elsewhere in the PR?

RoosterDragon replied:
The actors variable is now a plain enumerable - so I can't do that.

pchote commented:
I'd like to get this merged before the playtest to maximize our chances of finding any unexpected regressions.

penev92 commented:
Keep in mind that we do more than one playtest before a release. Some things will have to be left for the second playtest if we want to have a first one at all. This PR is no 2-minute work to review.

RoosterDragon commented:
Rebased and added the following changes:
- Change the actors list in World to a SortedSet, so the sync hashing in no longer relying on undefined behaviour (of the unordered HashSet).
- Added a RemoveAll method in world that simply passes through to the underlying RemoveAll method of the List<IEffect>. This prevents FlashTarget from needing to take an intermediate copy and is faster than incremental removes anyway.

Since this PR is a touch large, if anybody wants to just make a quick note of the things they don't like, I can always drop them or move them to another PR for easier reviewing.

obrakmann commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -37,7 +37,7 @@ public class Cargo : IPips, IIssueOrder, IResolveOrder, IOrderVoice, INotifyCrea
 		public readonly CargoInfo Info;
 		readonly Actor self;
-		readonly List<Actor> cargo = new List<Actor>();
+		readonly Queue<Actor> cargo = new Queue<Actor>();
```
I like this in principle. How about a `Stack` instead, though? You'd get that nice LIFO behaviour which would fit best for `Cargo`, I think.

pchote commented on OpenRA.Mods.Common/Traits/Player/PlaceBuilding.cs:
```diff
@@ -110,7 +110,7 @@ static void PlayBuildAnim(Actor self, ActorInfo unit)
 					.ToList();
 			var producer = producers.Where(x => x.Actor.IsPrimaryBuilding()).Concat(producers)
-				.FirstOrDefault();
+				.First();
```
Maybe keep this as FirstOrDefault, and then add a null check and early return?  We'll want this to be bulletproof against things dying in the same tick.

obrakmann replied:
`producer` ends up being a `TraitPair`, which is a struct, which can't be null (at least that's what Google told me when I wondered the same thing)

RoosterDragon replied:
Which would make keeping it as FirstOfDefault the correct thing to do then - nice catch and my bad.

obrakmann replied:
Yep, just got this:
```
Exception of type `System.InvalidOperationException`: Sequence contains no elements
  at System.Linq.Enumerable.First[TraitPair`1] (IEnumerable`1 source) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.RA.PlaceBuilding.PlayBuildAnim (OpenRA.Actor self, OpenRA.ActorInfo unit) [0x0007f] in /home/oliver/devel/openra/git/OpenRA.Mods.RA/Player/PlaceBuilding.cs:115
```

pchote commented:
Nothing sticks out to me as being potentially problematic, but I still haven't been able to test this.

obrakmann commented on OpenRA.Mods.Common/ServerTraits/MasterServerPinger.cs:
```diff
@@ -56,7 +56,7 @@ public void PingMasterServer(S server)
 			// by the main thread as clients join and leave.
 			var numPlayers = server.LobbyInfo.Clients.Where(c1 => c1.Bot == null && c1.Slot != null).Count();
-			var numBots = server.LobbyInfo.Clients.Where(c1 => c1.Bot != null).Count();
+			var numBots = server.LobbyInfo.Clients.Count - numPlayers;
```
I think this would count spectators as bots.

obrakmann commented:
Looks good overall. Didn't notice any regressions except for that one crash, and once that was fixed the game behaved normally. Comments need to be addressed, though.

RoosterDragon commented:
Rebased and fixed issues noted per comments.

Cargo now uses a stack which gives LIFO behaviour (which is a change from the previous queue like FIFO behaviour).

Mailaender commented on OpenRA.Mods.RA/Traits/Harvester.cs:
```diff
@@ -137,11 +137,8 @@ Actor ClosestProc(Actor self, Actor ignore)
 
-			// Reverse the found-path to find the refinery location instead of our location:
-			path.Reverse();
-
```
I assume these removals are an oversight while resolving rebase conflicts.

RoosterDragon replied:
Removing path.Reverse is intended. The code wants to look at the last element in the path and does so by reversing it and looking at the first element after that. Whilst that works, it's obviously more sensible to just directly look at the last element. The change a couple of lines down does that.

Unless you intended to mark another line?

Mailaender replied:
Thanks for the clarification. I thought `path` was re-used later and the harvester would otherwise not find it's way home.

obrakmann commented:
Looks good to me now. :+1:

Mailaender commented:
Also can't spot any regressions in the touched code. :+1:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/72399fc77b4cd4f3a9806308afdb62c752ed8e03...7b3ba5f6fe3641af9a54309c4af2a4dffbd9de86)
